routing/index_graph.cpp

```diff
@@ -13,27 +16,520 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+  m_roadIndex.AddJoint(fromFakeFtPoint, from);
+  m_jointIndex.AppendToJoint(from, fromFakeFtPoint);
+
+  return m_nextFakeFeatureId++;
+}
+
+uint32_t IndexGraph::AddFakeFeature(Joint::Id from, Joint::Id to,
+                                    vector<RoadPoint> const & geometrySource, double speed)
+{
+  CHECK_LESS(from, m_jointIndex.GetNumJoints(), ());
+  CHECK_LESS(to, m_jointIndex.GetNumJoints(), ());
+  CHECK_GREATER(geometrySource.size(), 1, ());
+
+  uint32_t fakeFeatureId = AddFakeLooseEndFeature(from, geometrySource, speed);
+  RoadPoint const toFakeFtPoint(fakeFeatureId, geometrySource.size() - 1 /* point id */);
+  m_roadIndex.AddJoint(toFakeFtPoint, to);
+  m_jointIndex.AppendToJoint(to, toFakeFtPoint);
+
+  return fakeFeatureId;
+}
+
+void IndexGraph::FindOneStepAsideRoadPoint(RoadPoint const & center, Joint::Id centerId,
+                                           vector<JointEdge> const & edges,
+                                           vector<Joint::Id> & oneStepAside) const
+{
+  oneStepAside.clear();
+  m_roadIndex.ForEachJoint(center.GetFeatureId(), [&](uint32_t /* pointId */, Joint::Id jointId) {
+    for (JointEdge const & e : edges)
+    {
+      if (e.GetTarget() == jointId)
+        oneStepAside.push_back(jointId);
+    }
+  });
+}
+
+bool IndexGraph::GetIngoingAndOutgoingEdges(Joint::Id centerId, bool graphWithoutRestrictions,
+                                            vector<JointEdge> & ingoingEdges,
+                                            vector<JointEdge> & outgoingEdges)
+{
+  ingoingEdges.clear();
+  outgoingEdges.clear();
+  GetEdgeList(centerId, false /* isOutgoing */, graphWithoutRestrictions, ingoingEdges);
+  if (ingoingEdges.empty())
+    return false;
+
+  GetEdgeList(centerId, true /* isOutgoing */, graphWithoutRestrictions, outgoingEdges);
+  if (outgoingEdges.empty())
+    return false;
+  return true;
+}
+
+bool IndexGraph::ApplyRestrictionPrepareData(RestrictionPoint const & restrictionPoint,
+                                             RestrictionInfo & restrictionInfo)
+{
+  vector<JointEdge> ingoingEdges;
+  vector<JointEdge> outgoingEdges;
+
+  GetEdgeList(restrictionPoint.m_centerId, false /* isOutgoing */, true /* graphWithoutRestrictions */, ingoingEdges);
+  vector<Joint::Id> fromOneStepAside;
+  FindOneStepAsideRoadPoint(restrictionPoint.m_from, restrictionPoint.m_centerId, ingoingEdges,
+                            fromOneStepAside);
+  if (fromOneStepAside.empty())
+    return false;
+
+  GetEdgeList(restrictionPoint.m_centerId, true /* isOutgoing */, true /* graphWithoutRestrictions */, outgoingEdges);
+  vector<Joint::Id> toOneStepAside;
+  FindOneStepAsideRoadPoint(restrictionPoint.m_to, restrictionPoint.m_centerId, outgoingEdges,
+                            toOneStepAside);
+  if (toOneStepAside.empty())
+    return false;
+
+  restrictionInfo.m_center = restrictionPoint.m_centerId;
+  restrictionInfo.m_from = fromOneStepAside.back();
```

Comment:
|fromOneStepAside| and |toOneStepAside| is filled with FindOneStepAsideRoadPoint() method. The idea of the method is to find a neighboring of restriction center joint which is belong to the ingoing(outgoing) feature. The thing is in very rare any feature of restriction could go through the restriction point several times. In that case I'll take any of them. There's a special notes near FindOneStepAsideRoadPoint():

  /// \note Taking into account the way of setting restrictions almost always |oneStepAside|
  /// will contain one or zero items. Meanwhile the it it's posible to draw map
  /// the (wrong) way |oneStepAside| will contain any number of items.